mplex/src/lib.rs

```diff
@@ -18,730 +18,326 @@
-    ) -> Self {
-        let name = name.into();
-
-        Substream {
-            id,
-            end,
-            name,
-            state,
-            buffer: None,
-        }
-    }
-
-    pub fn name(&self) -> Option<&Bytes> {
-        self.name.as_ref()
-    }
-
-    pub fn id(&self) -> u32 {
-        self.id
-    }
-}
-
-// TODO: We always zero the buffer, we should delegate to the inner stream.
-impl<T: AsyncRead, Buf: Array<Item = u8>> Read for Substream<T, Buf> {
-    fn read(&mut self, buf: &mut [u8]) -> io::Result<usize> {
-        let mut lock = match self.state.poll_lock() {
-            Async::Ready(lock) => lock,
-            Async::NotReady => return Err(io::ErrorKind::WouldBlock.into()),
-        };
-
-        read_stream(&mut lock, (self.id, buf))
+use bytes::Bytes;
+use core::{ConnectionUpgrade, Endpoint, StreamMuxer};
+use parking_lot::Mutex;
+use futures::prelude::*;
+use futures::{future, stream::Fuse, task};
+use tokio_io::{AsyncRead, AsyncWrite, codec::Framed};
+
+/// Configuration for the multiplexer.
+#[derive(Debug, Clone, Default)]
+pub struct MplexConfig;
+
+impl MplexConfig {
+    /// Builds the default configuration.
+    #[inline]
+    pub fn new() -> MplexConfig {
+        Default::default()
     }
 }
 
-impl<T: AsyncRead, Buf: Array<Item = u8>> AsyncRead for Substream<T, Buf> {}
-
-impl<T: AsyncWrite, Buf: Array> Write for Substream<T, Buf> {
-    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
-        let mut lock = match self.state.poll_lock() {
-            Async::Ready(lock) => lock,
-            Async::NotReady => return Err(io::ErrorKind::WouldBlock.into()),
-        };
-
-        let mut buffer = self.buffer
-            .take()
-            .unwrap_or_else(|| io::Cursor::new(buf_from_slice(buf)));
-
-        let out = write_stream(
-            &mut *lock,
-            write::WriteRequest::substream(MultiplexHeader::message(self.id, self.end)),
-            &mut buffer,
-        );
-
-        if buffer.position() < buffer.get_ref().len() as u64 {
-            self.buffer = Some(buffer);
-        }
-
-        out
-    }
+impl<C, Maf> ConnectionUpgrade<C, Maf> for MplexConfig
+where
+    C: AsyncRead + AsyncWrite,
+{
+    type Output = Multiplex<C>;
+    type MultiaddrFuture = Maf;
+    type Future = future::FutureResult<(Self::Output, Self::MultiaddrFuture), IoError>;
+    type UpgradeIdentifier = ();
+    type NamesIter = iter::Once<(Bytes, ())>;
 
-    fn flush(&mut self) -> io::Result<()> {
-        let mut lock = match self.state.poll_lock() {
-            Async::Ready(lock) => lock,
-            Async::NotReady => return Err(io::ErrorKind::WouldBlock.into()),
+    #[inline]
+    fn upgrade(self, i: C, _: (), endpoint: Endpoint, remote_addr: Maf) -> Self::Future {
+        let out = Multiplex {
+            inner: Arc::new(Mutex::new(MultiplexInner {
+                inner: i.framed(codec::Codec::new()).fuse(),
+                buffer: Vec::with_capacity(32),
```

Comment:
Instead of having a single shared buffer for all substreams have one for each? Otherwise the buffer may leak elements for substreams which have already gone.

Reply:
I've decided to keep a list of the opened substreams instead, ignore messages that belong to closed substreams, and remove the existing elements from the buffer when a substream is closed.